patches/nginx-1.13.6-resolver_conf_parsing.patch

```diff
@@ -0,0 +1,192 @@
++{
++    ngx_url_t                        u;
++    ngx_resolver_connection_t       *rec;
++    ngx_fd_t                         fd;
++    ngx_file_t                       file;
++    u_char                           buf[NGX_RESOLVER_FILE_BUF_SIZE];
++    ngx_uint_t                       address, j;
++    ssize_t                          n, i;
++    enum {
++        sw_nameserver,
++        sw_address,
++        sw_skip
++    } state;
++
++    file.name.data = path;
++    file.name.len = ngx_strlen(path);
++
++    if (ngx_conf_full_name(cf->cycle, &file.name, 1) != NGX_OK) {
++        return NGX_ERROR;
++    }
++
++    fd = ngx_open_file(file.name.data, NGX_FILE_RDONLY,
++                       NGX_FILE_OPEN, 0);
++
++    if (fd == NGX_INVALID_FILE) {
++        ngx_conf_log_error(NGX_LOG_EMERG, cf, ngx_errno,
++                           ngx_open_file_n " \"%s\" failed", file.name.data);
++
++        return NGX_ERROR;
++    }
++
++    ngx_memzero(&file, sizeof(ngx_file_t));
++
++    file.fd = fd;
++    file.name.data = path;
++    file.name.len = ngx_strlen(path) - 1;
++    file.log = cf->log;
++
++    state = sw_nameserver;
++
++    n = ngx_read_file(&file, buf, NGX_RESOLVER_FILE_BUF_SIZE, 0);
++
++    if (ngx_close_file(file.fd) == NGX_FILE_ERROR) {
++        ngx_conf_log_error(NGX_LOG_ALERT, cf, ngx_errno,
++                           ngx_close_file_n " \"%s\" failed", file.name.data);
++    }
++
++    if (n == NGX_ERROR) {
++        return NGX_ERROR;
++    }
++
++    if (n == 0) {
++        return NGX_OK;
++    }
++
++    for (i = 0; i < n; /* void */) {
++        if (buf[i] == '#' || buf[i] == ';') {
++            state = sw_skip;
++        }
++
++        switch (state) {
++
++        case sw_nameserver:
++
++            if ((size_t) n - i >= sizeof("nameserver ") - 1
++                && ngx_memcmp(buf + i, "nameserver ",
++                              sizeof("nameserver ") - 1) == 0)
++            {
++                state = sw_address;
++                address = 0;
++                i += sizeof("nameserver ") - 1;
++
++                continue;
++            }
++
++            break;
++
++        case sw_address:
++
++            if (address == 0) {
++                address = i;
++            }
++
++            if (buf[i] == LF || i == n - 1) {
++                ngx_memzero(&u, sizeof(ngx_url_t));
++
++                u.url.data = buf + address;
++                u.url.len = i - address;
++                u.default_port = 53;
++
++                if (ngx_parse_url(cf->pool, &u) != NGX_OK) {
++                    if (u.err) {
++                        ngx_conf_log_error(NGX_LOG_EMERG, cf, 0,
++                                           "%s in resolver \"%V\"",
++                                           u.err, &u.url);
++                    }
++
++                    return NGX_ERROR;
++                }
++
++                rec = ngx_array_push_n(&r->connections, u.naddrs);
```

Comment:
In hindsight, one of the things that I believe #247 is missing is the lack of test for whether a given IP address is already cached by the resolver or not.

Similarly in this case, do we care/are we doing anything to prevent the same nameserver from being added twice? Like:
```
# /etc/resolv.conf
nameserver 8.8.8.8

# nginx.conf
http {
    resolver 8.8.8.8;
    resolver local=on;
}
```

Reply:
That's interesting. AFAIK [you can't do multiple `resolver` directives within the same block](http://lxr.nginx.org/source/src/http/ngx_http_core_module.c#5112). However, NGINX does not prohibit you from doing:

```
resolver 8.8.8.8 8.8.8.8;
```

either. In the description of this PR, I described the behavior of `local` as:

> If local=on, then in addition to the resolvers specified, NGINX will parse /etc/resolv.conf and use nameservers found inside as resolvers.

It seems that allowing duplicates makes more sense? Plus de-duping this will be very painful due to how those data structure works currently.

Reply:
Yeah, sorry I think my example should have been:
```
# nginx.conf
http {
    resolver 8.8.8.8 local=on;
}
```

> It seems that allowing duplicates makes more sense? Plus de-duping this will be very painful due to how those data structure works currently.

Agreed - just raising the concern for us to be aware, but if nginx allows it, it is probably better to not try to add too much magic around it.